lib/rubocop/cop/style/double_pipe_equals.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for potential usage of the `||=` operator.
+      #
+      # @example
+      #   # bad
+      #   name = name ? name : 'Bozhidar'
+      #
+      #   # bad
+      #   name = 'Bozhidar' unless name
+      #
+      #   # good - set name to 'Bozhidar', only if it's nil or false
+      #   name ||= 'Bozhidar'
+      class DoublePipeEquals < Cop
+        MSG = 'Use the double pipe equals operator `||=` instead.'.freeze
+
+        def_node_matcher :ternary_assignment?, <<-PATTERN
+          (lvasgn _var
+            (if
+              (lvar _var)
+              (lvar _var)
+              _))
+        PATTERN
+
+        def_node_matcher :unless_assignment?, <<-PATTERN
+          (if
+            (lvar _var) nil
+            (lvasgn _var
+              _))
+        PATTERN
+
+        def on_if(node)
+          return unless unless_assignment?(node)
+          add_offense(node)
+        end
+
+        def on_lvasgn(node)
+          return unless ternary_assignment?(node)
+          add_offense(node)
+        end
+
+        private
+
+        def autocorrect(node)
+          if ternary_assignment?(node)
+            autocorrect_ternary(node)
+          else
+            autocorrect_unless(node)
+          end
+        end
+
+        def autocorrect_ternary(node)
+          lambda do |corrector|
+            corrector.replace(node.loc.operator, '||=')
+
+            _, right = *node
+            _, _, default = *right
+            corrector.replace(right.source_range, default.source)
+          end
+        end
+
+        def autocorrect_unless(node)
+          lambda do |corrector|
+            corrector.replace(node.loc.keyword, '||=')
+
+            first, _, last = *node
+            variable, default = *last
+            corrector.replace(last.source_range, variable.to_s)
+            corrector.replace(first.source_range, default.source)
+          end
+        end
```

Comment:
The auto-correction generates an invalid code if `unless` is not modifier.
For example: 


```ruby
x = nil
unless x
  x = 1
end
```


```bash
$ rubocop -D --only Style/DoublePipeEquals
Inspecting 1 file
C

Offenses:

test.rb:2:1: C: Style/DoublePipeEquals: Use the double pipe equals operator ||= instead.
unless x ...
^^^^^^^^

1 file inspected, 1 offense detected




$ rubocop -D --only Style/DoublePipeEquals -a
Inspecting 1 file
E

Offenses:

test.rb:2:1: E: Lint/Syntax: unexpected token tOP_ASGN
(Using Ruby 2.1 parser; configure using TargetRubyVersion parameter, under AllCops)
||= 1
^^^
test.rb:2:1: C: [Corrected] Style/DoublePipeEquals: Use the double pipe equals operator ||= instead.
unless x ...
^^^^^^^^
test.rb:4:1: E: Lint/Syntax: unexpected token kEND
(Using Ruby 2.1 parser; configure using TargetRubyVersion parameter, under AllCops)
end
^^^

1 file inspected, 3 offenses detected, 1 offense corrected




$ cat test.rb
x = nil
||= 1
  x
end
```